apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts

```diff
@@ -0,0 +1,1029 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import {
+  Colors,
+  JsonFile,
+  FileSystemStats,
+  IFileSystemCreateLinkOptions,
+  IColorableSequence
+} from '@rushstack/node-core-library';
+import * as crypto from 'crypto';
+import { Tslint, Typescript } from '@microsoft/rush-stack-compiler-3.7';
+import {
+  ExtendedTypescript,
+  IEmitResolver,
+  IExtendedEmitResult,
+  IEmitHost,
+  IEmitTransformers,
+  IExtendedProgram,
+  IExtendedSourceFile,
+  IResolveModuleNameResolutionHost
+} from './privateTypings/ExtendedTypescript';
+
+import { SubprocessRunnerBase } from '../../utilities/subprocess/SubprocessRunnerBase';
+import { Async } from '../../utilities/Async';
+import { resolvePackagePath } from '../../utilities/resolvePackage';
+import { IExtendedLinter } from './privateTypings/ExtendedTslint';
+import { IEmitModuleKindBase, ISharedTypescriptConfiguration } from '../../cli/actions/BuildAction';
+
+const ASYNC_LIMIT: number = 100;
+
+export interface ITypescriptBuilderConfiguration extends ISharedTypescriptConfiguration {
+  buildFolder: string;
+  typescriptToolPath: string;
+  tslintToolPath: string;
+  eslintToolPath: string;
+
+  lintingEnabled: boolean;
+
+  watchMode: boolean;
+
+  /**
+   * The path to the tsconfig file being built.
+   */
+  tsconfigPath: string;
+
+  /**
+   * The path to a tslint config file to use to lint the files being compiled, if one exists
+   */
+  tslintPath: string | undefined;
+
+  /**
+   * The path of project's build cache folder
+   */
+  buildCacheFolder: string;
+}
+
+interface ICachedEmitModuleKind<TModuleKind> extends IEmitModuleKindBase<TModuleKind> {
+  cacheOutFolderPath: string;
+  isPrimary: boolean;
+}
+
+interface IFileToWrite {
+  filePath: string;
+  data: string;
+}
+
+interface ITsLintCacheData {
+  cacheVersion: string;
+  fileVersions: [string, string][];
```

Comment:
Use a proper interface instead of `[string, string]`

Reply:
TSLint gives up a `Map<string, string>` and we can use `Array.from(...)` on it to get a `[string, string][]`. Changing this to an object will make the cache file larger and require looping through the `Map` entries to construct an array of objects, which is generally pointless work for a file that's never expected to be human-read.

Reply:
Can you document the data type then and give an example value?